python/ray/rllib/a3c/common.py

```diff
@@ -11,27 +11,41 @@ def discount(x, gamma):
     return scipy.signal.lfilter([1], [1, -gamma], x[::-1], axis=0)[::-1]
 
 
-def process_rollout(rollout, gamma, lambda_=1.0):
-    """Given a rollout, compute its returns and the advantage."""
-    batch_si = np.asarray(rollout.states)
-    batch_a = np.asarray(rollout.actions)
-    rewards = np.asarray(rollout.rewards)
-    vpred_t = np.asarray(rollout.values + [rollout.r])
-
-    rewards_plus_v = np.asarray(rollout.rewards + [rollout.r])
+def process_rollout(rollout, reward_filter, gamma, lambda_=1.0):
+    """Given a rollout, compute its returns and the advantage.
+
+    TODO(rliaw): generalize this"""
+    batch_si = np.asarray(rollout.data["state"])
+    batch_a = np.asarray(rollout.data["action"])
+    rewards = np.asarray(rollout.data["reward"])
+    vpred_t = np.asarray(rollout.data["value"] + [rollout.last_r])
+
+    rewards_plus_v = np.asarray(rollout.data["reward"] + [rollout.last_r])
     batch_r = discount(rewards_plus_v, gamma)[:-1]
     delta_t = rewards + gamma * vpred_t[1:] - vpred_t[:-1]
     # This formula for the advantage comes "Generalized Advantage Estimation":
     # https://arxiv.org/abs/1506.02438
     batch_adv = discount(delta_t, gamma * lambda_)
+    for i in range(batch_adv.shape[0]):
+        batch_adv[i] = reward_filter(batch_adv[i])
 
-    features = rollout.features[0]
-    return Batch(batch_si, batch_a, batch_adv, batch_r, rollout.terminal,
+    features = rollout.data["features"][0]
+    return Batch(batch_si, batch_a, batch_adv, batch_r, rollout.is_terminal(),
```

Comment:
this is still wip, but I'd like to replace all sample batches with a standard class https://github.com/ericl/ray/blob/evaluator-multi-gpu/python/ray/rllib/optimizers/sample_batch.py to better support pluggable optimizers.

Reply:
https://github.com/openai/baselines/blob/master/baselines/common/dataset.py

Check this out - this is a lot cleaner than what's going on right now imo